src/cpp/src/visual_language/cdpruner/kernel_builder.cpp

```diff
@@ -0,0 +1,559 @@
+    size_t batch_size = visual_shape[0];
+    size_t num_tokens = visual_shape[1];
+    size_t feature_dim = visual_shape[2];
+    size_t total_operations = batch_size * num_tokens * num_tokens;  // Dominant operation is N^2
+
+    if (relevance_scores.get_shape().size() != 2) {
+        throw std::invalid_argument("Relevance scores must be 2D tensor [B, N]");
+    }
+    auto relevance_shape = relevance_scores.get_shape();
+    if (m_config.pruning_debug_mode) {
+        std::cout << "[CDPruner] Input tensors: relevance_scores[" << relevance_shape[0] << ", " << relevance_shape[1]
+                  << "]" << std::endl;
+    }
+
+    // Check shape consistency
+    if (relevance_shape[0] != batch_size || relevance_shape[1] != num_tokens) {
+        throw std::invalid_argument(
+            "Visual features and relevance scores must have consistent batch size and token count");
+    }
+
+    // Performance timing for kernel building steps
+    auto kernel_build_start = std::chrono::high_resolution_clock::now();
+
+    // Step 1: L2 normalize visual features along the last dimension
+    // This is equivalent to: image_normalized = image_features / image_features.norm(dim=-1, keepdim=True)
+    auto normalize_start = std::chrono::high_resolution_clock::now();
+    ov::Tensor normalized_features = l2_normalize_features(visual_features);
+    auto normalize_end = std::chrono::high_resolution_clock::now();
+    auto normalize_duration = std::chrono::duration_cast<std::chrono::microseconds>(normalize_end - normalize_start);
+
+    // Step 2: Compute similarity matrix L = normalized_features @ normalized_features.T
+    // This gives us the base similarity matrix [B, N, N]
+    auto similarity_start = std::chrono::high_resolution_clock::now();
+    // ov::Tensor similarity_matrix = compute_similarity_matrix(normalized_features);
+    ov::Tensor similarity_matrix = compute_similarity_matrix_gpu(normalized_features);
+    auto similarity_end = std::chrono::high_resolution_clock::now();
+    auto similarity_duration = std::chrono::duration_cast<std::chrono::microseconds>(similarity_end - similarity_start);
+
+    // Step 3: Build conditional kernel matrix L̃ = diag(r) · L · diag(r)
+    // Following CDPruner: kernel = relevance.unsqueeze(2) * similarity * relevance.unsqueeze(1)
+    auto conditional_start = std::chrono::high_resolution_clock::now();
+    ov::Tensor conditional_kernel = build_conditional_kernel(similarity_matrix, relevance_scores);
+    auto conditional_end = std::chrono::high_resolution_clock::now();
+    auto conditional_duration =
+        std::chrono::duration_cast<std::chrono::microseconds>(conditional_end - conditional_start);
+
+    auto kernel_build_end = std::chrono::high_resolution_clock::now();
+    auto total_kernel_duration =
+        std::chrono::duration_cast<std::chrono::microseconds>(kernel_build_end - kernel_build_start);
+
+    // Print performance breakdown only in debug mode
+    if (m_config.pruning_debug_mode) {
```

Comment:
The extensive use of std::cout for performance logging and timing (e.g., chrono durations, operation counts) makes it harder to follow the core algorithm logic. Consider wrapping all debug prints and timing code behind a helper function so the main algorithm remains clean and readable.

Reply:
updated.